src/collection.js

```diff
@@ -604,6 +657,55 @@ export default class Collection {
 		} );
 	}
 
+	/**
+	 * For a given item to be added to the collection, it validates item's "id" property or generates
+	 * a new one if missing. If the validation was successful or a new id was generated,
+	 * the id is returned by the method. Otherwise, an error is thrown.
+	 *
+	 * @private
+	 * @param {Object} item An item to be added to the collection.
+	 */
+	_getItemIdBeforeAdding( item ) {
+		const idProperty = this._idProperty;
+		let itemId;
+
+		if ( ( idProperty in item ) ) {
+			itemId = item[ idProperty ];
+
+			if ( typeof itemId != 'string' ) {
+				/**
+				 * This item's id should be a string.
+				 *
+				 * @error collection-add-invalid-id
+				 * @param {Object} item The item being added to the collection.
+				 * @param {module:utils/collection~Collection} collection The collection the item is added to.
+				 */
+				throw new CKEditorError( 'collection-add-invalid-id', {
+					item,
+					collection: this,
+				} );
+			}
+
+			if ( this.get( itemId ) ) {
+				/**
+				 * This item already exists in the collection.
+				 *
+				 * @error collection-add-item-already-exists
+				 * @param {Object} item The item being added to the collection.
+				 * @param {module:utils/collection~Collection} collection The collection the item is added to.
+				 */
+				throw new CKEditorError( 'collection-add-item-already-exists', {
+					item,
+					collection: this
+				} );
+			}
+		} else {
+			item[ idProperty ] = itemId = uid();
```

Comment:
I'm not a fan of double assignments. Maybe move the `itemId` to the first `if` and always return `item[ idProperty ]`?